File: dotnet/src/Ode.Adapter/Stores.cs
using System.Text.Json.Nodes;

namespace Ode.Adapter;

/// <summary>One referral episode. Mirrors stores.Episode.</summary>
public sealed class Episode
{
    public string ReferralId { get; init; } = "";
    public string DirectMessageId { get; init; } = "";
    public string SubmissionSetId { get; init; } = "";
    public string SenderDirectAddress { get; init; } = "";
    public string RecipientDirectAddress { get; init; } = "";
    public string? TaskId { get; set; }
    public string? ServiceRequestId { get; set; }
    public string Status { get; set; } = "requested";
    // COW: granular dental business-status (see Config.CowBusinessStatus).
    public string? BusinessStatus { get; set; }
    // Which side opened this episode in the harness: "dental" or "medical".
    public string? InitiatedBy { get; set; }
    public string Created { get; } = DateTime.UtcNow.ToString("o");
    public List<string> History { get; } = new();
    // The FHIR the bridge wrote for this episode, so a harness "inbox" can read it
    // (in dry-run there is no live server to query). Newest last.
    public List<JsonObject> Inbox { get; } = new();

    public void Note(string msg) => History.Add($"{DateTime.UtcNow:o} {msg}");

    public void AddToInbox(IEnumerable<JsonObject>? resources)
    {
        foreach (var res in resources ?? Enumerable.Empty<JsonObject>())
        {
            var clean = new JsonObject();
            foreach (var kv in res)
                if (!kv.Key.StartsWith("_")) clean[kv.Key] = kv.Value?.DeepClone();
            Inbox.Add(clean);
        }
    }

    public JsonObject ToJson()
    {
        var hist = new JsonArray();
        foreach (var h in History) hist.Add(h);
        var inbox = new JsonArray();
        foreach (var r in Inbox) inbox.Add(JsonX.Clone(r));
        return new JsonObject
        {
            ["referral_id"] = ReferralId,
            ["direct_message_id"] = DirectMessageId,
            ["submission_set_id"] = SubmissionSetId,
            ["sender_direct_address"] = SenderDirectAddress,
            ["recipient_direct_address"] = RecipientDirectAddress,
            ["task_id"] = TaskId,
            ["service_request_id"] = ServiceRequestId,
            ["status"] = Status,
            ["business_status"] = BusinessStatus,
            ["initiated_by"] = InitiatedBy,
            ["created"] = Created,
            ["history"] = hist,
            ["inbox"] = inbox,
        };
    }
}

/// <summary>In-memory episode store. Subclass for a persistent store. Mirrors stores.CorrelationStore.</summary>
public class CorrelationStore
{
    private readonly Dictionary<string, Episode> _byReferral = new();

    public Episode Create(Episode ep) { _byReferral[ep.ReferralId] = ep; return ep; }
    public Episode? Get(string referralId) => _byReferral.TryGetValue(referralId, out var e) ? e : null;
    public IReadOnlyList<JsonObject> All() => _byReferral.Values.Select(e => e.ToJson()).ToList();
}

/// <summary>Direct address &lt;-&gt; ODE Native endpoint. Mirrors stores.Directory.</summary>
public class Directory
{
    private readonly Dictionary<string, string> _directToFhir;
    public Directory(Dictionary<string, string>? mappings = null) => _directToFhir = mappings ?? new();

    public string FhirEndpointFor(string directAddress) =>
        _directToFhir.TryGetValue(directAddress, out var v) ? v : Config.Settings.OdeNativeBaseUrl;
}

File: python/ode_adapter/stores.py
"""Correlation store + directory — core stateful services.

These are infrastructure the engine owns (not ports), but both are written behind
simple classes so they can be subclassed for a persistent store or a real provider
directory without touching the engine.
"""
from __future__ import annotations

from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone

from .config import settings


@dataclass
class Episode:
    referral_id: str
    direct_message_id: str
    submission_set_id: str
    sender_direct_address: str
    recipient_direct_address: str
    task_id: str | None = None
    service_request_id: str | None = None
    status: str = "requested"
    # COW: granular dental business-status (see config.COW_BUSINESS_STATUS).
    business_status: str | None = None
    # Which side opened this episode in the harness: "dental" or "medical".
    initiated_by: str | None = None
    created: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    history: list[str] = field(default_factory=list)
    # The FHIR the bridge wrote for this episode, so a harness "inbox" can read it
    # (in dry-run there is no live server to query). Newest last.
    inbox: list[dict] = field(default_factory=list)

    def note(self, msg: str) -> None:
        self.history.append(f"{datetime.now(timezone.utc).isoformat()} {msg}")

    def add_to_inbox(self, resources: list[dict]) -> None:
        for res in resources or []:
            self.inbox.append({k: v for k, v in res.items()
                               if not k.startswith("_")})


class CorrelationStore:
    """In-memory episode store. Subclass for SQLite/Postgres in production
    (episodes must outlive restarts and tolerate late/out-of-order messages)."""

    def __init__(self):
        self._by_referral: dict[str, Episode] = {}

    def create(self, ep: Episode) -> Episode:
        self._by_referral[ep.referral_id] = ep
        return ep

    def get(self, referral_id: str) -> Episode | None:
        return self._by_referral.get(referral_id)

    def all(self) -> list[dict]:
        return [asdict(e) for e in self._by_referral.values()]


class Directory:
    """Direct address <-> ODE Native endpoint / identity. Seed statically here;
    back with a real provider directory (DNS+LDAP / FHIR Endpoint) in production."""

    def __init__(self, mappings: dict[str, str] | None = None):
        self._direct_to_fhir = mappings or {}

    def fhir_endpoint_for(self, direct_address: str) -> str:
        return self._direct_to_fhir.get(direct_address, settings.ode_native_base_url)
